quinn-h3/src/connection.rs

```diff
@@ -16,60 +17,128 @@ use crate::{
     proto::{
         connection::{Connection, Error as ProtoError},
         frame::HttpFrame,
-        ErrorCode,
+        ErrorCode, StreamType,
     },
-    streams::{NewUni, RecvUni, SendControlStream},
+    streams::{NewUni, RecvUni},
     Error, Settings,
 };
 
-pub struct ConnectionDriver {
-    conn: ConnectionRef,
+pub struct ConnectionDriver(pub(crate) ConnectionRef);
+
+impl Future for ConnectionDriver {
+    type Output = Result<(), Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut conn = self.0.h3.lock().unwrap();
+
+        conn.poll_incoming_uni(cx)?;
+        conn.poll_send(cx)?;
+        conn.poll_recv_control(cx)?;
+        conn.poll_incoming_bi(cx)?;
+
+        if conn.inner.is_closing() && conn.inner.requests_in_flight() == 0 {
+            return Poll::Ready(Ok(()));
+        }
+        Poll::Pending
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct ConnectionRef {
+    pub h3: Arc<Mutex<ConnectionInner>>,
+    pub quic: quinn::Connection,
+}
+
+impl ConnectionRef {
+    pub fn new(
+        quic: quinn::Connection,
+        side: Side,
+        uni_streams: IncomingUniStreams,
+        bi_streams: IncomingBiStreams,
+        settings: Settings,
+    ) -> Result<Self, ProtoError> {
+        Ok(Self {
+            quic: quic.clone(),
```

Comment:
Does it really make sense to hold `quic` in the `ConnectionRef` and also the `ConnectionInner`? I'd think that one ought to be enough.

Reply:
I've been bothered by this choice at one moment but after checking again I just figured out something can be done in combination with the `set_error()` helper you proposed.